pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -93,83 +118,220 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
+	// then we need to promote without any considerations for deletion protection via machine hooks
+	// E.g scale up during bootstrap
+	if numVotingMembers < c.mastersReplicaCount {
+		return true, nil
+	}
+
+	// TODO: Once the machine hooks are reconciled, check the preconditions for deletion hooks
+	// before promoting to a max of N+1 voting members.
+	// For now we just blindly promote all learners
+	return true, nil
+}
+
+// TODO: Copied as is from ClusterMemberRemoval controller, consolidate in pkg/operator/ceohelpers
+// findMachineByNodeInternalIP finds machine that matches the given nodeInternalIP
+// is safe because the MAO:
+//  sync the addresses in the Machine with those assigned to real nodes by the cloud provider ,
+//  checks that the Machine and Node lists match before issuing a serving certification for Kubelet
+//  when the host disappears from the cloud side, it stops updating the Machine so the addresses and information should persist there as a tombstone as the Machine is marked Failed
+func (c *ClusterMemberController) findMachineByNodeInternalIP(nodeInternalIP string) (*machinev1beta1.Machine, error) {
+	machines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return nil, err
+	}
+	for _, machine := range machines {
+		for _, addr := range machine.Status.Addresses {
+			if addr.Type == corev1.NodeInternalIP {
+				if addr.Address == nodeInternalIP {
+					return machine, nil
+				}
+			}
 		}
 	}
 	return nil, nil
 }
 
-// getValidPodFQDNToScale goes through the list on unready pods and
-// returns a resolvable  podFQDN. If none of the DNSes are available
-// yet it will return collected errors.
-func (c *ClusterMemberController) getEtcdPeerHostToScale(podToAdd *corev1.Pod) (string, error) {
-	network, err := c.networkLister.Get("cluster")
+func (c *ClusterMemberController) getPeerURLForNode(node *corev1.Node) (string, error) {
+	internalIP, err := c.getInternalIPForNode(node)
 	if err != nil {
 		return "", err
 	}
-	node, err := c.nodeLister.Get(podToAdd.Spec.NodeName)
+	return fmt.Sprintf("https://%s:2380", internalIP), nil
+}
+
+func (c *ClusterMemberController) getInternalIPForNode(node *corev1.Node) (string, error) {
+	network, err := c.networkLister.Get("cluster")
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to list cluster network: %w", err)
+	}
+
+	internalIP, err := dnshelpers.GetEscapedPreferredInternalIPAddressForNodeName(network, node)
+	if err != nil {
+		return "", fmt.Errorf("failed to get escaped preferred internal IP for node: %w", err)
 	}
+	return internalIP, nil
+}
 
-	return dnshelpers.GetEscapedPreferredInternalIPAddressForNodeName(network, node)
+func isURLMappedToMember(peerURL string, members []*etcdserverpb.Member) bool {
+	for _, m := range members {
+		if peerURL == m.PeerURLs[0] {
+			return true
+		}
+	}
+	return false
+}
+
+func countVotingMembers(members []*etcdserverpb.Member) int {
```

Comment:
do we want to count unhealthy voting members in here?  does it vary by call site?